**If you lost points on the last checkpoint you can get them back by responding to TA/IA feedback**  

Update/change the relevant sections where you lost those points, make sure you respond on GitHub Issues to your TA/IA to call their attention to the changes you made here.

Please update your Timeline... no battle plan survives contact with the enemy, so make sure we understand how your plans have changed.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

# COGS 108 - EDA Checkpoint

# Names

- Levy Sahoo
- Keenan Serrao
- Leela Stuepfert
- Maya Ammar

# Research Question

-  How do socioeconomic factors influence a nation's performance at the Olympic games?
    - What specific factors are the most signifcant towards a nation's medal performance? 



## Background and Prior Work

The Olympic games have been deemed as notably the most prestigious sporting event in human history, it’s a collective movement where all the best athletes in the world assemble together for one thing in mind: Glory for their nation. The way to achieve this is to attain a medal. Although it’s fair to say that the Olympic games are a stage that gathers the best of the best on the same platform to evaluate their performances in their respective events, we need to reconsider if there’s actually more to what meets the eye test. Instead of simply asking who will find a spot on the podium, seeing the journeys of the athletes with various backgrounds shifts the focus of the question to how did they arrive to the opportunity to compete for a shot at the podium, and are there quantitative features that we can use to assess regarding their country of origin and socioeconomic backgrounds to help us answer the begging question: 

**Do economic factors such as GDP and population count correlate to a nation’s successive medal performance at the Olympic games?** 

**“What do other research studies and deem about factors correlated to a nation’s Olympic Success?”**

A 2000 study<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_ref-1) conducted by economists Andrew Bernard and Meghan Busse, explored how economic factors such as GDP and population size impact a nation’s Olympic success. The resport<a name="cite_ref-2"></a>[<sup>2</sup>](#cite_ref-2) summarized by Michael Klien, Professor of International Economic Affairs at Tufts University, identified real GDP as the most significant predictor of a country's medal count, estimating that a 10% increase in income per capita yields a 6.9% increase in medals, assuming population remains constant. Similarly, a 10% population increase, while holding income steady, leads to a 3.6% rise in medals. These findings suggest a clear correlation between economic scale and Olympic performance, providing a foundation for the relevance of these variables in predicting success. On the contrary, some researchers dispute the validity of these regressors upon a nation’s successive medal performance, a more recent 2022 study<a name="cite_ref-3"></a>[<sup>3</sup>](#cite_ref-3) questions the significance of GDP and population on Olympic outcomes, finding that variables such as GDP size, corruption ranking, athlete count, and topography don’t significantly impact medal standings. Instead, factors like inflation rates, economic activity, and income classification seem to offer alternative perspectives on predicting Olympic success. This work introduces a broader range of variables, which can contribute to refining model regressors and understanding potential limitations of traditional economic measures. Since some of these features aren’t open to public availability, such metrics can deliver insights towards the interpretation of the error terms of the linear regression models used to analyze our objective. A separate Georgia Tech analysis<a name="cite_ref-4"></a>[<sup>4</sup>](#cite_ref-4) identifies other influential factors, such as country size and healthcare expenditure per capita, in determining Olympic performance. By highlighting healthcare investment as a predictor, it expands beyond purely economic indicators, suggesting that a nation’s healthcare infrastructure may play a vital role in supporting elite athletic development.

While each study offers valuable insights, they present varying perspectives on the significance of GDP and population as predictors of Olympic success, it’s not definitive. Researchers may emphasize GDP and population as core predictors, while others cast doubt on their significance, proposing additional variables such as inflation and economic activity. Georgia Tech’s emphasis on healthcare investment introduces yet another dimension. This divergence highlights the complexity of determining Olympic success factors and suggests that GDP and population, though potentially impactful, may not be exhaustive predictors. By integrating these varied viewpoints, our project can better assess the significance of socioeconomic regressors with common data available as well as analyze the combined effect of these and other regressors, using econometric modeling to explore their potential in predicting Olympic medal ranking success while considering possible model limitations and interpreting the error sources.

1. <a name="cite_note-1"></a> [^](#cite_ref-1) Bernard, A. (1 Dec 2002) Who Wins the Olympic Games: Economic Resources and Medal Totals *Review of Economics and Statistics*. https://faculty.tuck.dartmouth.edu/images/uploads/faculty/andrew-bernard/olymp60restat_finaljournalversion.pdf 
2. <a name="cite_note-2"></a> [^](#cite_ref-2) Klein, M. (17 Jul 2024) What Determines Countries’ Olympic Success? *ECONOFACT* https://econofact.org/what-determines-countries-olympic-success
3. <a name="cite_note-3"></a> [^](#cite_ref-3) Sasha, W. (28 May 2022) Assessment of Olympic performance in relation to economic, demographic, geographic, and social factors: quantile and Tobit approaches *Taylor & Francis Online*. https://www.tandfonline.com/doi/full/10.1080/1331677X.2022.2080735#abstract
3. <a name="cite_note-4"></a> [^](#cite_ref-4) Boudreau, J. The Miracle on Thin Ice: How A Nation's GDP Affects its Olympic Performance *Georgia Tech University*. https://repository.gatech.edu/server/api/core/bitstreams/1aa2b537-c3de-4177-8295-3fcd3a03a965/content#:~:text=We%20estimate%20that%20GDP%20per,bronze%20medals%20a%20country%20receives


# Hypotheses


***Null hypothesis (H<sub>0</sub>):*** There is no significant correlation between a country’s `GDP per capita` or `population` size and its Olympic medal perfomance. 


- Indication that there is no evidence to suggest that ***GDP per capita*** and ***population*** sizes are affiliated the number of medals a country earns and their overall medal performance.


***Alternative hypothesis (H<sub>a</sub>):*** There is a significant correlation between a country’s `GDP per capita` and `population` size upon its Olympic medal performance. 


- There exists evidence that ***higher GDP per capitas*** and ***larger population sizes*** are likely to be correlated with a nation's olympic medal performance, as increased population and resources may correlate to national team size and likelihood of podium success. 



***Predictive Reasoning:*** We believe the answer is lies in the alternative hypothesis, that wealthier and more populus countries experience greater Olympic Success. Due to unobserved variables that allow First-World nations to allocate a greater investment towards athletic development, niche sport specific training, and funding, we can expect these nations to achieve greater success in the Olympics. Additionally, countries with a larger population have a larger population of potential athletes to select from, influencing the size of a nation's olympic team, increasing the likelihood of finding more athletes on the podium. Rejecting the Null hypothesis will allow us to accept the alternative and help us verify whether socioeconomic and demographic advantages contribute to Olympic success or not. 




## Overview of Data(sets)


#### **Dataset #1: Summer Olympic Medals 1896 - 2020**
  - Source: https://www.kaggle.com/datasets/ramontanoeiro/summer-olympic-medals-1986-2020 
  - Number of observations: 1344
  - Number of variables: 8 

"Summer Olympic Medals 1896–2020," is a comprehensive historical record of medals awarded in every Summer Olympic Games from 1896 to 2020, also sourced from Kaggle. Each row provides data on a specific country’s performance in a given Olympic year.

Dataset **columns** include:

- `Year`
- `Host_country`
- `Host_city`
- `Country_Name`
- `Country_Code`
- `Gold`
- `Silver`
- `Bronze`

This dataset contains thousands of observations spanning numerous Olympic editions, enabling us to analyze trends in Olympic performance over time. While this dataset does not include economic indicators, it will serve as a foundation for historical analysis of country specific medal achievements, which can later be compared with economic data for a more in depth exploration of trends.

#### **Dataset #2: GDP by Country 1999 - 2022**

  - Source: https://www.kaggle.com/datasets/alejopaullier/-gdp-by-country-1999-2022
  - Number of observations: 180
  - Number of variables: 24

"GDP by Country 1999–2022," offers annual GDP data for all countries worldwide, covering the years:
`1999`
...
`2022` 
and the GDPs measured in billions of US dollars. It is structured with rows for each `country` and columns for each `year`, which makes it ideal for tracking economic growth and fluctuations over time. Each cell represents a `country`’s GDP for a specific `year`, allowing for both cross-country comparisons within a single year and longitudinal analysis within a single `country` over multiple years. When this data is merged with the Olympic medal datasets, it will allow us to investigate if economic factors, such as GDP growth or decline, correlate with Olympic success across different time periods.

Dataset **columns** include:

- `Country`

Range of 23 Year columns from:
- `1999` - `2022`


#### **Dataset #3: 2024 Olympic Medals vs GDP**
  - Source: https://www.kaggle.com/datasets/ernestitus/2024-olympics-medals-vs-gdp
  - Number of observations: 90
  - Number of variables: 10

"2024 Olympics: Medals vs GDP," is sourced from Kaggle and modified from Mohamed Yosef’s “2024 Olympics Medals and Economic Status.” It provides data on the performance and economic indicators of countries participating in the 2024 Olympics. This dataset includes information for 90 countries, with each row representing a country and columns detailing attributes such as:

- `country`
- `country_code` (Nation's Abreviation)
- `region` (e.g., Europe, Asia). 

Olympic performance is recorded through **medal counts**: 

- `gold`
- `silver`
- `bronze`
- `total`

Economic data fields include `gdp`, `gdp_year` (latest GDP data), and `population`, allowing for a thorough exploration of potential correlations between a `country`’s economic profile and its Olympic performance.

## Summer Olympic Medals 1896 - 2020

#### Cleaning Process Explanation: 

In data set we identified missing values using `df.isnull().sum()` and determined that the missing entries in non-critical columns could be safely ignored without impacting the analysis. Rows with missing values in essential columns (`Country`, `Year`, and `Medal`) were dropped using `df.dropna(subset=['Country', 'Year', 'Medal'])`. To ensure that there was consistency across datasets, we standardized country names using a mapping dictionary (e.g., `{'United States': 'USA', 'Great Britain': 'UK'}`). This mapping was applied with `df['Country'].replace(mapping_dict)`. The Medal column was transformed by creating binary columns for Gold, Silver, and Bronze. This allowed for aggregating medal counts more efficiently. The data was then grouped by Country and Year, summing up medals with `df.groupby(['Country', 'Year']).sum()`. This provided a view of total medals won by each country per Olympic year, which is relevant for our research question. It is important to note that for easier analysis, log-transforming medal counts were considered if they displayed skewed distributions, given that extreme values (from countries with high medal counts) could affect correlation results. Finally, to remain parallel with data from the GDPs Dataset, we filter `medals_cleaned`to consider `[medals_cleaned['Year'] >= 1999]`

Start by loading the dataset and taking a look at the first few rows as well as the column names and types.


In [971]:
medals = pd.read_csv('Summer_olympic_Medals.csv')
medals.head()

Year Host_country Host_city   Country_Name Country_Code  Gold  Silver  \
0  1896       Greece    Athens  Great Britain          GBR     2       3   
1  1896       Greece    Athens        Hungary          HUN     2       1   
2  1896       Greece    Athens         France          FRA     5       4   
3  1896       Greece    Athens  United States          USA    11       7   
4  1896       Greece    Athens        Germany          GER     6       5   

   Bronze  
0       2  
1       3  
2       2  
3       2  
4       2

In [972]:
medals.dtypes

Year             int64
Host_country    object
Host_city       object
Country_Name    object
Country_Code    object
Gold             int64
Silver           int64
Bronze           int64
dtype: object

Next, look for null values in the dataset

In [973]:
# Count number of nans in each column
medals.isnull().sum()

Year             0
Host_country     0
Host_city        0
Country_Name     0
Country_Code    86
Gold             0
Silver           0
Bronze           0
dtype: int64

In [974]:
# Display rows with nans
medals[medals.isnull().any(axis=1)].head()

Year Host_country       Host_city  Country_Name Country_Code  Gold  \
1165  2016       Brazil  Rio de Janeiro       Denmark          NaN     2   
1166  2016       Brazil  Rio de Janeiro     Argentina          NaN     3   
1167  2016       Brazil  Rio de Janeiro        Sweden          NaN     2   
1168  2016       Brazil  Rio de Janeiro       Ukraine          NaN     2   
1169  2016       Brazil  Rio de Janeiro  South Africa          NaN     2   

      Silver  Bronze  
1165       6       7  
1166       1       0  
1167       6       3  
1168       5       4  
1169       6       2

As we can see below, we only have null country codes from the year 2016. This can be fixed by filling in the values with country codes from previous years.

In [975]:
medals[medals.isnull().any(axis=1)]['Year'].unique()

array([2016])

Fill in null country codes with values from previous years

In [976]:
country_code_map = dict(zip(medals['Country_Name'], medals['Country_Code']))
medals_cleaned = medals.assign(Country_Code = medals['Country_Name'].map(country_code_map).fillna(medals['Country_Code']))

Create a new column called `Total_Medals` that sums the total number of medals won by each country in each year.

In [977]:
medals_cleaned['Total_Medals'] = medals_cleaned['Gold'] + medals_cleaned['Silver'] + medals_cleaned['Bronze']

Recheck for null values

In [978]:
medals_cleaned.isnull().sum()

Year             0
Host_country     0
Host_city        0
Country_Name     0
Country_Code    11
Gold             0
Silver           0
Bronze           0
Total_Medals     0
dtype: int64

In [979]:
# Display rows with nans
medals_cleaned[medals_cleaned.isnull().any(axis=1)].head()['Country_Name'].unique()

array(['North Korea', 'United Arab Emirates', 'Russia', 'Niger',
       'Burundi'], dtype=object)

In [980]:
manual_country_codes = {
    'North Korea': 'PRK',
    'United Arab Emirates': 'UAE',
    'Russia': 'RUS',
    'Niger': 'NIG',
    'Burundi': 'BDI',
    'Trinidad and Tobago': 'TTO',
    'Vietnam': 'VIE',
    'Independent Olympic Athletes': 'IOA',
    'Tajikistan': 'TJK',
    'Algeria': 'ALG',
    'Singapore': 'SGP'
}


# Apply these manual country codes to the data
medals_cleaned['Country_Code'] = medals_cleaned.apply(
    lambda row: manual_country_codes.get(row['Country_Name'], row['Country_Code']), axis=1
)

In [981]:
medals_cleaned.isnull().sum()

Year            0
Host_country    0
Host_city       0
Country_Name    0
Country_Code    0
Gold            0
Silver          0
Bronze          0
Total_Medals    0
dtype: int64

Our data is now clean and ready to use

In [982]:
medals_cleaned.head()

Year Host_country Host_city   Country_Name Country_Code  Gold  Silver  \
0  1896       Greece    Athens  Great Britain          GBR     2       3   
1  1896       Greece    Athens        Hungary          HUN     2       1   
2  1896       Greece    Athens         France          FRA     5       4   
3  1896       Greece    Athens  United States          USA    11       7   
4  1896       Greece    Athens        Germany          GER     6       5   

   Bronze  Total_Medals  
0       2             7  
1       3             6  
2       2            11  
3       2            20  
4       2            13

In [983]:
medals.shape

(1344, 8)

In [984]:
medals_cleaned = medals_cleaned[medals_cleaned['Year'] >= 1999]
medals_cleaned

Year Host_country Host_city Country_Name Country_Code  Gold  Silver  \
838   2000    Australia    Sydney        Spain          ESP     3       3   
839   2000    Australia    Sydney       Canada          CAN     3       3   
840   2000    Australia    Sydney         Iran          IRI     3       0   
841   2000    Australia    Sydney       Turkey          TUR     3       0   
842   2000    Australia    Sydney      Belarus          BLR     3       3   
...    ...          ...       ...          ...          ...   ...     ...   
1339  2020        Japan     Tokyo         Fiji          FIJ     1       0   
1340  2020        Japan     Tokyo      Estonia          EST     1       0   
1341  2020        Japan     Tokyo       Latvia          LAT     1       0   
1342  2020        Japan     Tokyo      Bermuda          BER     1       0   
1343  2020        Japan     Tokyo     Thailand          THA     1       0   

      Bronze  Total_Medals  
838        5            11  
839        8            14  
840        1             4  
841        2             5  
842       11            17  
...      ...           ...  
1339       1             2  
1340       1             2  
1341       1             2  
1342       0             1  
1343       1             2  

[506 rows x 9 columns]

## GDP by Country 1999 - 2022

#### Data Cleaning Explanation:

For this dataset, the data was filtered to include only Olympic `year`s (e.g., `2000, 2004, 2008`, etc.) using `df[df['Year'].isin([2000, 2004, 2008, 2012, 2016, 2020])]`. This alignment ensures that the GDP data directly corresponds with Olympic medal data. Using df.isna().all, we see that this dataset does have 0 values. One of the challenges we addressed was interpolating these 'empty' GDPs, so we'll attempt to use polynomial or spline interpolation techniques to replace 0s with estimates that fit the GDP trends. Another challenge was removing the excessive formality in the `country` names such as 'Republic', 'Democratic', and other `country` identification terminology. We did this to align the `Country` records with the `medals_cleaned` data and used `df['Country'].replace(mapping_dict)`. This will help us to merge datasets later without issues for modeling. 

In [1829]:
org_hist_gdps = pd.read_csv('GDPs99-22.csv')
org_hist_gdps.isna().all().head()

Country    False
1999       False
2000       False
2001       False
2002       False
dtype: bool

In [1830]:
org_hist_gdps

Country    1999     2000     2001    2002    2003    2004  \
0    Afghanistan, Rep. of.       0        0        0   4.084   4.585   5.971   
1                  Albania   3.444    3.695    4.096   4.456     5.6   7.452   
2                  Algeria  48.845   54.749   55.181  57.053  68.013  85.016   
3                   Angola   6.153    9.135    8.936  11.386  13.956    19.8   
4      Antigua and Barbuda   0.652    0.678     0.71   0.718   0.754   0.818   
..                     ...     ...      ...      ...     ...     ...     ...   
175              Venezuela  97.977  117.153  122.872  92.889  83.442   112.8   
176                Vietnam  28.684   31.196   32.504  35.148   39.63  45.548   
177     Yemen, Republic of    7.53    9.561    9.533   9.985  11.869  13.565   
178                 Zambia   3.132    3.238     3.64   3.775   4.326    5.44   
179               Zimbabwe   5.963    8.136   12.883  30.856  10.515     4.7   

        2005     2006     2007  ...     2013     2014     2015     2016  \
0      7.309    8.399    9.892  ...   21.555   24.304        0        0   
1      8.376    9.133   10.163  ...    14.91   16.053   11.591   12.204   
2     102.38  114.322  116.158  ...  190.432  203.449  175.077   181.71   
3     30.632   43.759    55.37  ...  136.415  151.089  102.011   98.815   
4      0.875    0.962    1.026  ...    1.404    1.494    1.285    1.328   
..       ...      ...      ...  ...      ...      ...      ...      ...   
175  143.443  181.608  219.372  ...  403.123  409.562  131.855  133.534   
176   53.053   60.995   68.298  ...  135.729  148.914  198.805   214.75   
177   15.193     18.7   21.657  ...   40.003   42.687        0        0   
178    7.271   10.942   10.104  ...   21.829   23.613   24.466   25.158   
179    4.552     5.54   18.281  ...    6.214    6.851   13.905   14.292   

        2017     2018     2019     2020     2021     2022  
0          0        0        0        0        0        0  
1     13.214   14.341   15.553   16.996    16.77   18.012  
2    192.256  202.179  210.906   219.16  163.812  168.195  
3    105.369  112.533  119.403   127.15   70.339   74.953  
4      1.386    1.458    1.536    1.617    1.405    1.534  
..       ...      ...      ...      ...      ...      ...  
175  144.227  155.096   170.41  184.364   44.893   43.546  
176  229.845  247.415  265.987  287.257  368.002  415.493  
177        0        0        0        0        0        0  
178    27.17   29.911   32.957   36.316   21.699   23.967  
179   14.994   15.736   16.609   17.529   25.791   27.806  

[180 rows x 24 columns]

In [1797]:
index_hist_gdps = org_hist_gdps.set_index('Country')
org_hist_gdps = index_hist_gdps[[year for year in index_hist_gdps.columns if int(year) % 4 == 0]].reset_index()

Some countries have formal titles and terminology that are not parallel with names in the other datasets:
- Look at Afghanistan and Yemen
- Both these nations have a form of the word "Republic" in their name

In [1798]:
org_hist_gdps['Country']

0      Afghanistan, Rep. of.
1                    Albania
2                    Algeria
3                     Angola
4        Antigua and Barbuda
               ...          
175                Venezuela
176                  Vietnam
177       Yemen, Republic of
178                   Zambia
179                 Zimbabwe
Name: Country, Length: 180, dtype: object

In [1799]:
## Make a Copy for Reference, we'll refer back to the Original CSV version later
hist_gdps = org_hist_gdps.copy()

In [1800]:
remove_punctuation = r'[.,\'-]'
hist_gdps['Country'] = hist_gdps['Country'].str.replace(remove_punctuation, '', regex=True)

remove_words = r'\b(Rep|Republic|Demo|Democratic|of|the|People|Côte d|Equatorial|Islamic|Former)\b'
hist_gdps['Country'] = hist_gdps['Country'].str.replace(remove_words, '', regex=True, case=False)

hist_gdps['Country'] = hist_gdps['Country'].str.strip()
hist_gdps

Country     2000    2004     2008     2012     2016     2020
0            Afghanistan        0   5.971   11.513   19.248        0        0
1                Albania    3.695   7.452   11.131   13.808   12.204   16.996
2                Algeria   54.749  85.016  126.889   177.83   181.71   219.16
3                 Angola    9.135    19.8   67.608  118.426   98.815   127.15
4    Antigua and Barbuda    0.678   0.818    1.074    1.322    1.328    1.617
..                   ...      ...     ...      ...      ...      ...      ...
175            Venezuela  117.153   112.8  231.959  394.106  133.534  184.364
176              Vietnam   31.196  45.548   76.414  123.505   214.75  287.257
177                Yemen    9.561  13.565   24.504   37.153        0        0
178               Zambia    3.238    5.44   10.519   20.208   25.158   36.316
179             Zimbabwe    8.136     4.7   63.039    5.528   14.292   17.529

[180 rows x 7 columns]

This is how many countries were changed as a result of the regex transformation

In [1801]:
transformation_differences = org_hist_gdps['Country'] != hist_gdps['Country']
num_differences = transformation_differences.sum()
num_differences

23

In [1802]:
different_records = hist_gdps[transformation_differences]
different_records

Country     2000     2004     2008     2012     2016  \
0                   Afghanistan        0    5.971   11.513   19.248        0   
10                      Bahamas    5.004    5.661    7.026    7.786        0   
31              Central African    0.962    1.309    1.745    2.428     1.84   
37                        Congo    4.303    6.539    9.954   14.525        0   
38                        Congo     3.22    4.349    6.878   13.186        0   
40                 Côte dIvoire   10.448   15.501   20.225   28.002   34.284   
43                        Czech   56.717  108.214  172.504  229.512  189.982   
47                    Dominican   20.059   18.435   33.843   51.474   69.908   
51                       Guinea    1.231     4.85    9.836   15.172   10.211   
59                       Gambia    0.421    0.401    0.606    0.835        0   
67                 GuineaBissau    0.216     0.27    0.362    0.515    1.116   
76                         Iran    96.44  161.261  241.006  403.811        0   
88                       Kyrgyz    1.368    2.215    3.646     5.54    7.659   
89                  Lao Peoples    1.735    2.508    4.448    7.069        0   
97          Macedonia  Yugoslav    3.583    5.377    7.548   10.946        0   
141                      Slovak   20.374   42.015   78.303  109.056   90.544   
147          St Kitts and Nevis     0.33    0.406    0.557      0.6     0.93   
148                    St Lucia    0.685    0.763    1.084    1.119    1.498   
149  St Vincent and  Grenadines    0.335    0.408    0.555    0.733    0.794   
155                 Syrian Arab   19.861   24.703   36.815   74.323        0   
156      Taiwan Province  China  321.374  322.299  389.991  443.539  540.073   
160              TimorLeste Dem    0.316    0.339     0.53    0.967        0   
177                       Yemen    9.561   13.565   24.504   37.153        0   

        2020  
0          0  
10         0  
31     2.916  
37         0  
38         0  
40     50.86  
43   206.851  
47    85.005  
51    10.856  
59         0  
67     1.563  
76         0  
88     10.43  
89         0  
97         0  
141  115.111  
147    1.117  
148    1.767  
149    0.944  
155        0  
156  650.902  
160        0  
177        0

Now let's check to see which countries still have formal terminology in the next few cells. We also need to find countries that have historically participated in the Olympics. To do this, we'll find them from `medals_cleaned` countries, because we only need countries that have participated in the Olympics

In [1804]:
valid_countries = set(medals_cleaned['Country_Name'].unique())
valid_countries

{'Afghanistan',
 'Algeria',
 'Argentina',
 'Armenia',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bahrain',
 'Barbados',
 'Belarus',
 'Belgium',
 'Bermuda',
 'Botswana',
 'Brazil',
 'Bulgaria',
 'Burkina Faso',
 'Burundi',
 'Cameroon',
 'Canada',
 'Chile',
 'China',
 'Chinese Taipei',
 'Colombia',
 'Costa Rica',
 'Croatia',
 'Cuba',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Dominican Republic',
 'Ecuador',
 'Egypt',
 'Eritrea',
 'Estonia',
 'Ethiopia',
 'Fiji',
 'Finland',
 'France',
 'Gabon',
 'Georgia',
 'Germany',
 'Ghana',
 'Great Britain',
 'Greece',
 'Grenada',
 'Guatemala',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'Independent Olympic Athletes',
 'India',
 'Indonesia',
 'Iran',
 'Ireland',
 'Israel',
 'Italy',
 'Ivory Coast',
 'Jamaica',
 'Japan',
 'Jordan',
 'Kazakhstan',
 'Kenya',
 'Kosovo',
 'Kuwait',
 'Kyrgyzstan',
 'Latvia',
 'Lithuania',
 'Macedonia',
 'Malaysia',
 'Mauritius',
 'Mexico',
 'Moldova',
 'Mongolia',
 'Montenegro',
 'Morocco',
 'Mozambique',
 'Namib

In [1770]:
valid_countries_num = medals_cleaned['Country_Name'].nunique()
valid_countries_num

128

So there are 128 Countries in the `medals_cleaned` Dataset, but do we have all 128 in our cleaned filtered_gdps?

In [1771]:
filtered_gdps = hist_gdps[hist_gdps['Country'].isin(valid_countries)]
filtered_gdps

Country     2000     2004     2008      2012      2016      2020
0    Afghanistan        0    5.971   11.513    19.248         0         0
2        Algeria   54.749   85.016  126.889    177.83    181.71    219.16
5      Argentina  284.204  151.958  277.912   315.832   578.692   633.889
6        Armenia    1.912    3.573    8.579     9.086    10.672    12.939
7      Australia  390.017  639.356  854.072  1,039.69  1,253.00  1,515.70
..           ...      ...      ...      ...       ...       ...       ...
172      Uruguay   20.086   13.268   22.211    47.355    57.151    70.107
173   Uzbekistan   13.717   12.001   21.727    44.892    68.699    88.986
175    Venezuela  117.153    112.8  231.959   394.106   133.534   184.364
176      Vietnam   31.196   45.548   76.414   123.505    214.75   287.257
179     Zimbabwe    8.136      4.7   63.039     5.528    14.292    17.529

[105 rows x 7 columns]

No we do not! To find the countries we need to add in which are currently `missing_countries`, we'll subtract `filtered_gdps`'s Countries from `valid_countries` (from `medals_cleaned`)

In [1772]:
filtered_countries = set(filtered_gdps['Country'])
missing_countries = set(valid_countries - filtered_countries)
missing_countries

{'Bermuda',
 'Chinese Taipei',
 'Cuba',
 'Czech Republic',
 'Dominican Republic',
 'Great Britain',
 'Hong Kong',
 'Independent Olympic Athletes',
 'Ivory Coast',
 'Kosovo',
 'Kyrgyzstan',
 'Macedonia',
 'Montenegro',
 'North Korea',
 'North Macedonia',
 'Puerto Rico',
 'ROC',
 'San Marino',
 'Serbia and Montenegro',
 'Slovakia',
 'South Korea',
 'Syria',
 'Yugoslavia'}

Now we know which countries are missing, let's add their rows to filtered_gdps
- They're missing because they have formal names and we haven't transformed them regular expressions. 
- we need to find the rows of the missing nations from `org_hist_gdps` (if they're available)
- some of these Countries from `medals_cleaned` aren't available `org_hist_gdps`

In [1773]:
org_hist_gdps_replace = org_hist_gdps.copy()

In [1774]:
org_hist_gdps_replace['Country'] = org_hist_gdps_replace['Country'].replace({
    'Taiwan Province of China': 'Chinese Taipei', 
    'Czech Republic': 'Czech Republic', 
    'Dominican Republic': 'Dominican Republic', 
    'United Kingdom': 'Great Britain', 
    'Hong Kong SAR': 'Hong Kong', 
    "Côte d'Ivoire": 'Ivory Coast', 
    'Kyrgyz Republic': 'Kyrgyzstan', 
    'Macedonia, Former Yugoslav Republic of': 'North Macedonia', 
    'Korea': 'South Korea', 
    'Serbia': 'Serbia and Montenegro', 
    'Slovak Republic': 'Slovakia', 
    'Syrian Arab Republic': 'Syria'
    })

In [1775]:
org_hist_gdps_replace[org_hist_gdps_replace['Country'] == "Ivory Coast"]

Country    2000    2004    2008    2012    2016   2020
40  Ivory Coast  10.448  15.501  20.225  28.002  34.284  50.86

Not all the 23 missing countries that were in `missing_countries` are available in the `org_hist_gdps` dataset. This is fine, we still have a handful of 12 countries to work with. Since 105 countries are already in `filtered_gdps`, we just need to add the 12 countries to `filtered_gdps`, therefore there should be **117** rows in `filtered_gdps` when cleaning is finished. 

In [1776]:
manual_country_names = {
    'Taiwan Province of China': 'Chinese Taipei', 
    'Czech Republic': 'Czech Republic', 
    'Dominican Republic': 'Dominican Republic', 
    'United Kingdom': 'Great Britain', 
    'Hong Kong SAR': 'Hong Kong', 
    "Côte d'Ivoire'": 'Ivory Coast', 
    'Kyrgyz Republic': 'Kyrgyzstan', 
    'Macedonia, Former Yugoslav Republic of': 'North Macedonia', 
    'Korea': 'South Korea', 
    'Serbia': 'Serbia and Montenegro', 
    'Slovak Republic': 'Slovakia', 
    'Syrian Arab Republic': 'Syria'
    }

We find the rows of the `missing_countries` we have GDP data about

In [1777]:
manual_country_values = set(manual_country_names.values())
manual_filtered_rows = org_hist_gdps_replace[org_hist_gdps_replace['Country'].isin(manual_country_values)]
manual_filtered_rows

Country      2000      2004      2008      2012      2016  \
40             Ivory Coast    10.448    15.501    20.225    28.002    34.284   
43          Czech Republic    56.717   108.214   172.504   229.512   189.982   
47      Dominican Republic    20.059    18.435    33.843    51.474    69.908   
71               Hong Kong   168.754   165.823   213.915   243.899   322.166   
86             South Korea   511.961   681.227  1,000.53  1,005.74  1,450.05   
88              Kyrgyzstan     1.368     2.215     3.646      5.54     7.659   
97         North Macedonia     3.583     5.377     7.548    10.946         0   
137  Serbia and Montenegro     8.963    24.518    38.516    51.111     37.74   
141               Slovakia    20.374    42.015    78.303   109.056    90.544   
155                  Syria    19.861    24.703    36.815    74.323         0   
156         Chinese Taipei   321.374   322.299   389.991   443.539   540.073   
170          Great Britain  1,445.19  2,155.16  2,822.66  2,611.27  3,054.84   

         2020  
40      50.86  
43    206.851  
47     85.005  
71    405.781  
86   1,898.76  
88      10.43  
97          0  
137    50.031  
141   115.111  
155         0  
156   650.902  
170  3,851.98

In [1778]:
manual_filtered_rows.shape[0]

12

In [1779]:
filtered_gdps_before = filtered_gdps.copy()
filtered_gdps_before.shape[0]


105

Finally, we have *117* countries that we have GDP Data on and have participated in the Olympics historically:

In [1780]:
filtered_gdps = pd.concat([filtered_gdps_before, manual_filtered_rows], ignore_index=True)
gdps_cleaned = filtered_gdps.copy()
gdps_cleaned

Country      2000      2004      2008      2012      2016  \
0              Afghanistan         0     5.971    11.513    19.248         0   
1                  Algeria    54.749    85.016   126.889    177.83    181.71   
2                Argentina   284.204   151.958   277.912   315.832   578.692   
3                  Armenia     1.912     3.573     8.579     9.086    10.672   
4                Australia   390.017   639.356   854.072  1,039.69  1,253.00   
..                     ...       ...       ...       ...       ...       ...   
112  Serbia and Montenegro     8.963    24.518    38.516    51.111     37.74   
113               Slovakia    20.374    42.015    78.303   109.056    90.544   
114                  Syria    19.861    24.703    36.815    74.323         0   
115         Chinese Taipei   321.374   322.299   389.991   443.539   540.073   
116          Great Britain  1,445.19  2,155.16  2,822.66  2,611.27  3,054.84   

         2020  
0           0  
1      219.16  
2     633.889  
3      12.939  
4    1,515.70  
..        ...  
112    50.031  
113   115.111  
114         0  
115   650.902  
116  3,851.98  

[117 rows x 7 columns]

### **^^^WE NEED TO FIND A WAY TO REPLACE AND INTERPOLATE ZEROS, THEY MUST FIT GDP TRENDS** ###

Just to make sure that all the countries in `gdps_cleaned`

In [1781]:
all_valid = filtered_gdps['Country'].isin(valid_countries).all()
all_valid

True

Countries that participated in the Olympics but we DON'T have GDP Data on, we can ignore these

In [1782]:
filtered_countries = set(filtered_gdps['Country'])
missing_countries = set(valid_countries - filtered_countries)
missing_countries

{'Bermuda',
 'Cuba',
 'Independent Olympic Athletes',
 'Kosovo',
 'Macedonia',
 'Montenegro',
 'North Korea',
 'Puerto Rico',
 'ROC',
 'San Marino',
 'Yugoslavia'}

## Dataset #3: 2024 Olympic Medals vs GDP

#### Data Cleaning Explanation:

We will address this dataset later for merging purposes. As of now, there appears to be no null data from `olympics24.isna().all()`. A conversion metric for `GDP` was needed to convert the total GDP of each country into billions USD. Finally GDP will have to be readjusted for '2024' instead of 2023 as listed in `GDP_Year`. 

In [1876]:
olympics24 = pd.read_csv('olympics.csv')
olympics24.isna().all()

/Users/levsah/miniforge3/lib/python3.10/site-packages/IPython/core/displayhook.py:281: UserWarning: Output cache limit (currently 1000 entries) hit.
Flushing oldest 200 entries.
  warn('Output cache limit (currently {sz} entries) hit.\n'


country         False
country_code    False
region          False
gold            False
silver          False
bronze          False
total           False
gdp             False
gdp_year        False
population      False
dtype: bool

In [1877]:
olympics24

country country_code         region  gold  silver  bronze  total  \
0   United States          USA  North America    40      44      42    126   
1           China          CHN           Asia    40      27      24     91   
2           Japan          JPN           Asia    20      12      13     45   
3       Australia          AUS        Oceania    18      19      16     53   
4          France          FRA         Europe    16      26      22     64   
..            ...          ...            ...   ...     ...     ...    ...   
85           Peru          PER  South America     0       0       1      1   
86          Qatar          QAT           Asia     0       0       1      1   
87      Singapore          SGP           Asia     0       0       1      1   
88       Slovakia          SVK         Europe     0       0       1      1   
89         Zambia          ZMB         Africa     0       0       1      1   

         gdp  gdp_year  population  
0   81695.19      2023       334.9  
1   12614.06      2023      1410.7  
2   33834.39      2023       124.5  
3   64711.77      2023        26.6  
4   44460.82      2023        68.2  
..       ...       ...         ...  
85   7789.87      2023        34.4  
86  87480.42      2022         2.7  
87  84734.26      2023         5.9  
88  24470.24      2023         5.4  
89   1369.13      2023        20.6  

[90 rows x 10 columns]

Let's see if the countries from the 2024 olympics are in `valid_countries`

In [1878]:
invalid_countries = set(olympics24['country']) - valid_countries
invalid_countries

{'Albania',
 'Cape Verde',
 'Dominica',
 'Pakistan',
 'Peru',
 'St Lucia',
 'Taiwan',
 'Zambia'}

Let's remove these countries from `Olympics24` since there is no record from `medals_cleaned` of these countries

In [1879]:
olympics24 = olympics24[~olympics24['country'].isin(invalid_countries)]
olympics24

country country_code         region  gold  silver  bronze  total  \
0   United States          USA  North America    40      44      42    126   
1           China          CHN           Asia    40      27      24     91   
2           Japan          JPN           Asia    20      12      13     45   
3       Australia          AUS        Oceania    18      19      16     53   
4          France          FRA         Europe    16      26      22     64   
..            ...          ...            ...   ...     ...     ...    ...   
82    Puerto Rico          PRI      Caribbean     0       0       2      2   
84    Ivory Coast          CIV         Africa     0       0       1      1   
86          Qatar          QAT           Asia     0       0       1      1   
87      Singapore          SGP           Asia     0       0       1      1   
88       Slovakia          SVK         Europe     0       0       1      1   

         gdp  gdp_year  population  
0   81695.19      2023       334.9  
1   12614.06      2023      1410.7  
2   33834.39      2023       124.5  
3   64711.77      2023        26.6  
4   44460.82      2023        68.2  
..       ...       ...         ...  
82  36779.06      2023         3.2  
84   2728.80      2023        28.9  
86  87480.42      2022         2.7  
87  84734.26      2023         5.9  
88  24470.24      2023         5.4  

[82 rows x 10 columns]

We need to scale the GDP of the countries into billions, taking `gdp` and scaling it by 0.359 to get the relative GDP in billions USD

In [1882]:
olympics24.rename(columns={'gdp': 'gdp_per_capita'}, inplace=True)
olympics24['gdp_total'] = ((olympics24['gdp_per_capita'] * olympics24['population']) / 1000).round(2)



/var/folders/t5/1y7x3kc535bctp93r0gt23rm0000gn/T/ipykernel_34359/2939918307.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  olympics24.rename(columns={'gdp': 'gdp_per_capita'}, inplace=True)
/var/folders/t5/1y7x3kc535bctp93r0gt23rm0000gn/T/ipykernel_34359/2939918307.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  olympics24['gdp_total'] = ((olympics24['gdp_per_capita'] * olympics24['population']) / 1000).round(2)


In [1883]:
olympics24

country country_code         region  gold  silver  bronze  total  \
0   United States          USA  North America    40      44      42    126   
1           China          CHN           Asia    40      27      24     91   
2           Japan          JPN           Asia    20      12      13     45   
3       Australia          AUS        Oceania    18      19      16     53   
4          France          FRA         Europe    16      26      22     64   
..            ...          ...            ...   ...     ...     ...    ...   
82    Puerto Rico          PRI      Caribbean     0       0       2      2   
84    Ivory Coast          CIV         Africa     0       0       1      1   
86          Qatar          QAT           Asia     0       0       1      1   
87      Singapore          SGP           Asia     0       0       1      1   
88       Slovakia          SVK         Europe     0       0       1      1   

    gdp_per_capita  gdp_year  population  gdp_total  
0         81695.19      2023       334.9   27359.72  
1         12614.06      2023      1410.7   17794.65  
2         33834.39      2023       124.5    4212.38  
3         64711.77      2023        26.6    1721.33  
4         44460.82      2023        68.2    3032.23  
..             ...       ...         ...        ...  
82        36779.06      2023         3.2     117.69  
84         2728.80      2023        28.9      78.86  
86        87480.42      2022         2.7     236.20  
87        84734.26      2023         5.9     499.93  
88        24470.24      2023         5.4     132.14  

[82 rows x 11 columns]

# Results

## Exploratory Data Analysis

Carry out whatever EDA you need to for your project.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

### Section 1 of EDA - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [1794]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [ ]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

# Ethics & Privacy

With regards to privacy, terms of use and security, the dataset being used is publicly available on Kaggle, and there are no direct privacy issues involving individuals. However, it is important to note that the data is used in accordance with Kaggle's terms of service, and that any findings are not misrepresented or used to make harmful generalizations about certain countries or groups.

There exist potential biases and limitations in the data set present through the exclusion of certain nations from the set. Countries that have not participated in the 2024 Olympics or those that have missing GDPs are excluded from this set; which in turn creates a bias by limiting the analysis to only those countries for which their data is available. As such, this leads to a possibility that the data may be disproportionate or skewed to negatively affect less economically developed nations, as their absence could lead to underestimating their performance potential of nations with fewer economic resources. Moreover, analyzing whether GDP correlates with Olympic medal count implicitly assumes that economic power should or does lead to success in sports. This assumption can marginalize less wealthy nations as GDP alone is not an adequate representation of a country's ability to win medals. There are many other cultural and social factors that play significant roles in this case; which is why it is essential to acknowledge that athletic talent is universal, but opportunities to develop that talent are not. The framing of our findings must be approached with caution, emphasizing that the purpose is to identify patterns rather than to justify disparities between nations. 

Our group will acknowledge the potential biases and limitations in the dataset through the discussion of our results, which ensures that these potential biases are clear to readers. When communicating this analysis, we will be transparent about the notion that correlations observed do not imply causation, and we will emphasize that our analysis is intended to explore patterns rather than make normative claims about countries’ athletic and economic capabilities. Additionally, in order to mitigate these biases we plan to include other factors that influence GDP such as government spending on sports, total investment, etc. so as to gather a more nuanced conclusion with regards to the determinants of Olympic success. Furthermore, throughout the data cleaning and model representation process, we will assess the distribution of countries based on GDP and medals to identify any biases in representation; by evaluating model performance for potential overfitting to high-GDP countries. This is significant to the reporting phase, as we consider how our results create a general conclusion, which will be framed in order to avoid the implication that economic power alone determines success. For instance, while we may observe a correlation between GDP and medal count, it is critical to stress that this does not mean only wealthy countries can succeed in the Olympics. Which in turn dives into the ethical implications of this research question, more specifically how these findings may potentially reflect global inequalities in sports development; by raising awareness of the socioeconomic factors that contribute to Olympic success.

# Team Expectations 


Read over the [COGS108 Team Policies](https://github.com/COGS108/Projects/blob/master/COGS108_TeamPolicies.md) individually. Then, include your group’s expectations of one another for successful completion of your COGS108 project below. Discuss and agree on what all of your expectations are. Discuss how your team will communicate throughout the quarter and consider how you will communicate respectfully should conflicts arise. By including each member’s name above and by adding their name to the submission, you are indicating that you have read the COGS108 Team Policies, accept your team’s expectations below, and have every intention to fulfill them. These expectations are for your team’s use and benefit — they won’t be graded for their details.

* *Team Expectation 1*
* *Team Expectation 2*
* *Team Expecation 3*
* ...

# Project Timeline Proposal

Specify your team's specific project timeline. An example timeline has been provided. Changes the dates, times, names, and details to fit your group's plan.

If you think you will need any special resources or training outside what we have covered in COGS 108 to solve your problem, then your proposal should state these clearly. For example, if you have selected a problem that involves implementing multiple neural networks, please state this so we can make sure you know what you’re doing and so we can point you to resources you will need to implement your project. Note that you are not required to use outside methods.



| Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 1/20  |  1 PM | Read & Think about COGS 108 expectations; brainstorm topics/questions  | Determine best form of communication; Discuss and decide on final project topic; discuss hypothesis; begin background research | 
| 1/26  |  10 AM |  Do background research on topic | Discuss ideal dataset(s) and ethics; draft project proposal | 
| 2/1  | 10 AM  | Edit, finalize, and submit proposal; Search for datasets  | Discuss Wrangling and possible analytical approaches; Assign group members to lead each specific part   |
| 2/14  | 6 PM  | Import & Wrangle Data (Ant Man); EDA (Hulk) | Review/Edit wrangling/EDA; Discuss Analysis Plan   |
| 2/23  | 12 PM  | Finalize wrangling/EDA; Begin Analysis (Iron Man; Thor) | Discuss/edit Analysis; Complete project check-in |
| 3/13  | 12 PM  | Complete analysis; Draft results/conclusion/discussion (Wasp)| Discuss/edit full project |
| 3/20  | Before 11:59 PM  | NA | Turn in Final Project & Group Project Surveys |